ion/bitstream.go

```diff
@@ -396,26 +396,45 @@ func (b *bitstream) ReadAnnotationIDs() ([]uint64, error) {
 		panic("not an annotation")
 	}
 
-	lengthValue, lengthOfLength, err := b.readVarUintLen(b.len)
+	annotFieldLength, lengthOfAnnotFieldLength, err := b.readVarUintLen(b.len)
 	if err != nil {
 		return nil, err
 	}
 
-	if b.len-lengthOfLength <= lengthValue {
+	if annotFieldLength == 0 {
+		// An annotation with zero length is illegal because at least one annotation must be present.
+		return nil, &SyntaxError{"malformed annotation: at least one annotation must be specified",
+			b.pos - lengthOfAnnotFieldLength}
+	}
+
+	remainingAnnotationLength := b.len - lengthOfAnnotFieldLength - annotFieldLength
+
+	if remainingAnnotationLength <= 0 {
 		// The size of the annotations is larger than the remaining free space inside the
 		// annotation container.
-		return nil, &SyntaxError{"malformed annotation", b.pos - lengthOfLength}
+		return nil, &SyntaxError{"malformed annotation", b.pos - lengthOfAnnotFieldLength}
 	}
 
 	var as []uint64
-	for lengthValue > 0 {
-		id, idlen, err := b.readVarUintLen(lengthValue)
+	for annotFieldLength > 0 {
+		id, idlen, err := b.readVarUintLen(annotFieldLength)
 		if err != nil {
 			return nil, err
 		}
 
 		as = append(as, id)
-		lengthValue -= idlen
+		annotFieldLength -= idlen
+	}
+
+	// Get the raw data for the value enclosed within the annotation without advancing the reader.
+	annotatedData, err := b.in.Peek(int(remainingAnnotationLength))
```

Comment:
This could plausibly be quite large. Is there any way to do the needed checks without reading the whole annotated value into memory at once?

Reply:
Addressed. Now we're just peeking at the next byte one at a time.